### PR TITLE
FIX Avoid creating repository when it exists on remote

### PR DESCRIPTION
**Pull request by merveenoyan**
This PR partially fixes https://github.com/huggingface/huggingface_hub/issues/672. 

I observed that the following error pops up when I create a repository using huggingface API `create_repo` and then have a local repository cloning from the URL **using a token** ([here](https://github.com/huggingface/huggingface_hub/issues/672#issuecomment-1150972003))
`ValueError: No space_sdk provided. create_repo expects space_sdk to be one of ['gradio', 'streamlit', 'static'] when repo_type is 'space'`
So I noticed the following block was causing the problem:

```
if token is not None:
                whoami_info = self.client.whoami(token)
                user = whoami_info["name"]
                valid_organisations = [org["name"] for org in whoami_info["orgs"]]

                if namespace is not None:
                    repo_id = f"{namespace}/{repo_id}"
                repo_url += repo_id

                scheme = urlparse(repo_url).scheme
                repo_url = repo_url.replace(f"{scheme}://", f"{scheme}://user:{token}@")

                if namespace == user or namespace in valid_organisations:
                    self.client.create_repo(
                        repo_id=repo_id,
                        token=token,
                        repo_type=self.repo_type,
                        exist_ok=True,
                        private=self.private,
                    )
            else:
                if namespace is not None:
                    repo_url += f"{namespace}/"
                repo_url += repo_id
```

Only name being valid shouldn't be enough to create a repository, we have to check if the repository exists on remote so I added a small check on that (thanks @muellerzr for pointing out to particular HfApi function that does it).

With this, `space_sdk` error should be gone. I'll add test once @LysandreJik approves my logic in here.

Also this is separate but IDK if it's good to have an attribute `clone_from` and the `clone_from()` function. I can refactor couple of things I saw there that's not good to me as well.

### CODE REVIEW

File: src/huggingface_hub/repository.py
Comment (adrinjalali) [inline src/huggingface_hub/repository.py:650]: I'm very confused, why would `clone_from` create a repo remotely under any circumstances?

**Review by adrinjalali — COMMENTED** [review]
Also tests? :D

File: src/huggingface_hub/repository.py
Comment (merveenoyan) [reply src/huggingface_hub/repository.py:650]: It's because link given to `clone_from` might not be valid. Maybe it's better if I put a warning saying the remote repository doesn't exist instead. :)

Comment (merveenoyan) [reply src/huggingface_hub/repository.py:650]: or do we want to not create at all and just raise a warning?

Comment (merveenoyan) [reply src/huggingface_hub/repository.py]: I removed repo creation and added a warning.

Comment (merveenoyan) [reply src/huggingface_hub/repository.py]: AH OK

Comment (merveenoyan) [reply src/huggingface_hub/repository.py:650]: I just realized there are many tests that will fail and we'll break backwards compatibility if we change behavior of `clone_from`, I also thought that it was only used to clone existing repository when creating repository on local. pinging @LysandreJik about why the design is that way 😅

**Comment by merveenoyan** [comment]
_Super weird thing_
I can do following:
```
api = HfApi()

temp_repo_url = api.create_repo(
            repo_id="merve/temp", token=token, 
            repo_type="space", space_sdk="gradio"
        )
repo = Repository("./dummy", clone_from=temp_repo_url,
        use_auth_token=token
        )
```

But when I run the test I've written (below) I still get the same error. 
```
def test_init_clone_from(self):
        temp_repo_url = self._api.create_repo(
            repo_id=f"{self.REPO_NAME}-temp", token=self._token, 
            repo_type="space", space_sdk="gradio"
        )
        repo = Repository(WORKING_REPO_DIR, clone_from=temp_repo_url,
        use_auth_token=self._token
        )
        self._api.delete_repo(
                repo_id=f"valid_org/{self.REPO_NAME}", token=self._token
            )
```

I have other stuff to do so will leave this for monday.

File: src/huggingface_hub/repository.py
Comment (merveenoyan) [reply src/huggingface_hub/repository.py:650]: (ok apparently you left a comment but I didn't see it bcs I didn't refresh the page lol)

**Comment by merveenoyan** [comment]
@osanseviero yup I'm sure (I uninstalled and installed again)
What I'm testing here is that it should bypass creating the repository remotely given there's a repository remotely already (which we check with `HfApi.repo_info()`)

File: src/huggingface_hub/repository.py
Comment (julien-c) [reply src/huggingface_hub/repository.py:650]: (i agree that Repository clone_from shouldn't create a repo)

Comment (julien-c) [reply src/huggingface_hub/repository.py:650]: (can't find at the moment the past discussion with a potential explicit `Repository.create() -> Repository`)

Comment (LysandreJik) [reply src/huggingface_hub/repository.py:650]: Yes, I also think that `clone_from` shouldn't create the repository if it doesn't exist, it isn't really its place to do so. I wouldn't say removing it is a bugfix as when I implemented it I shared it as such, but we can deprecate it for sure.

**Comment by merveenoyan** [comment]
@osanseviero it passed 😏

**Comment by osanseviero** [comment]
This seems to be passing the tests now 😄

File: src/huggingface_hub/repository.py
Comment (merveenoyan) [reply src/huggingface_hub/repository.py:651]: I say we merge this and then deprecate it.

**Review by adrinjalali — APPROVED** [review]
I'm happy to merge and deprecate in a separate PR.

File: src/huggingface_hub/repository.py
Comment (LysandreJik) [reply src/huggingface_hub/repository.py:651]: SGTM!

File: src/huggingface_hub/repository.py
Comment (LysandreJik) [inline src/huggingface_hub/repository.py:656]: Wouldn't this also need a `space_sdk` in case the repository in question is a space? i.e., if the repo is a space that doesn't exist, it'll try to create it but fail at doing so.

As we're deprecating, maybe that's intended and we keep that use-case as it is (i.e., not working)

**Review by LysandreJik — COMMENTED** [review]
Looks good to me! Only left a comment regarding `space_sdk`.

File: src/huggingface_hub/repository.py
Comment (merveenoyan) [reply src/huggingface_hub/repository.py:656]: I will raise a `ValueError` in case someone tries to create Space that doesn't exist because otherwise I have to add `space_sdk` as a class attribute to `Repository` (we're later deprecating the repository creation in `clone_from` anyway so there's no point to add that). I'll add two tests, one for model repo and one for Space repo to see if the error is raised.  We can merge tomorrow ^^

Comment (LysandreJik) [reply src/huggingface_hub/repository.py:656]: Sounds great, thank you @merveenoyan!

**Comment by merveenoyan** [comment]
Please do not review I know what the problem with tests is :D

**Comment by merveenoyan** [comment]
I added three tests:
- `test_init_clone_from()` checks that when space exists it's clone well
- `test_clone_from_space()` makes sure the space is not created if it doesn't exist
- `test_clone_from_model()` makes sure that model is created if it doesn't exist.

**Comment by merveenoyan** [comment]
One of the tests are not passing because moon-landing staging is down 😅

**Comment by LysandreJik** [comment]
@merveenoyan it seems you're using an outdated version of `black`. Could you updated to the one required by `huggingface_hub` and run the code quality fix? Should be something like the following, from the root of your clone:

```
pip install -U -e .[quality]
make style
make quality
```

@SBrandeis, @XciD, would you have why we're getting errors 500 on the `test_init_clone_from` test? It seems like creating a space using `create_repo` doesn't work and throws an error 500. Reproducible code example:

```py
temp_repo_url = create_repo(
    repo_id=f"temp",
    token=token,
    repo_type="space",
    space_sdk="streamlit"
)
```

I tried with both `streamlit`, `gradio`. However, `static` seems to work just fine!

**Comment by christophe-rannou** [comment]
@LysandreJik I just created a repo using the CLI with success. Maybe there's an issue with your token ?

**Comment by LysandreJik** [comment]
@christophe-rannou the token is that of the CI, we're using the same token for all CI. Were you on the production endpoint or the staging endpoint?

**Comment by christophe-rannou** [comment]
Production one, it's the staging one that is not working ?

**Comment by XciD** [comment]
> Production one, it's the staging one that is not working ?

Seems related to CI, will investigate.

**Comment by merveenoyan** [comment]
This one fails `HfApiEndpointsTest.test_create_update_and_delete_space_repo` with 500 like it was failing for my PR, see below where the error comes from (due to picking other SDKs for others)

```
for sdk in SPACES_SDK_TYPES:
            SPACE_REPO_NAME = space_repo_name(sdk)
            self._api.create_repo(
                repo_id=SPACE_REPO_NAME,
                token=self._token,
                repo_type=REPO_TYPE_SPACE,
                space_sdk=sdk,
            )
            res = self._api.update_repo_visibility(
                repo_id=SPACE_REPO_NAME,
                token=self._token,
                private=True,
                repo_type=REPO_TYPE_SPACE,
            )
            self.assertTrue(res["private"])
            res = self._api.update_repo_visibility(
                repo_id=SPACE_REPO_NAME,
                token=self._token,
                private=False,
                repo_type=REPO_TYPE_SPACE,
            )
```

**Comment by LysandreJik** [comment]
This exact test fails on `main` since https://github.com/huggingface/huggingface_hub/pull/898 was merged, I would offer to skip that test until this is resolved, and merge this PR. Ok for you @merveenoyan ?

**Comment by merveenoyan** [comment]
@LysandreJik sure!

**Comment by osanseviero** [comment]
Was this supposed to be merged? As @LysandreJik mentioned, I think we should add the flag/decorator to skip the test so the tests are not failing (` @unittest.SkipTest` for example)

**Comment by LysandreJik** [comment]
No problem @osanseviero, I opened a PR to fix the problem there: https://github.com/huggingface/huggingface_hub/pull/908

But now we have another problem with our CI because of the new datasets release, so waiting until that is fixed to merge it.

**Comment by XciD** [comment]
About the spaces bug, we found it. 
Waiting for https://github.com/huggingface/moon-landing/pull/3252/commits/f2e8068348e225ba1f07c5d27a38843253e23dcb to be merged.

**Comment by XciD** [comment]
> About the spaces bug, we found it. Waiting for [huggingface/moon-landing@f2e8068](https://github.com/huggingface/moon-landing/commit/f2e8068348e225ba1f07c5d27a38843253e23dcb) to be merged.

Fixed

**Pull request merged**
